### PR TITLE
api: Support combination of condition and factor for ConditionalDimension

### LINKED ISSUE #2273: Conflict in time_M when using saved buffer and conditional dimension
When using a conditional dimension that contains `Ge` / `Le` types of conditions and a saved buffer defined over that conditional dimension, there is a conflict in the calculation of `time_M` for different objects:

```
from devito import *
from devito.symbolics import CondEq


grid = Grid(shape=(10, 10))
time = grid.time_dim

nt = 200
bounds = (10, 100)
factor = 5

condition = sympy.And(CondEq(time % 7, 0),
                      Ge(time, bounds[0]),
                      Le(time, bounds[1]), )

time_under = ConditionalDimension(name='timeu',
                                  parent=time,
                                  factor=7,
                                  condition=condition)

buffer_size = (bounds[1] - bounds[0] + factor) // factor + 1

rec = SparseTimeFunction(name='rec', grid=grid, npoint=1, nt=nt)

u = TimeFunction(name='u', grid=grid, space_order=2)
usaved = TimeFunction(name='usaved', grid=grid, space_order=2, time_dim=time_under, save=buffer_size)

eq = [Eq(usaved, u)] + rec.interpolate(u)

op = Operator(eq)
op(time_m=1, time_M=nt-1, usaved=usaved, rec=rec)
```
produces the error `ValueError: Unable to find unique value for key time_M, candidates: [199, range(1, 139)]`.

### PR DESCRIPTION
**Pull request by mloubout**
Fixes #2273

### CODE REVIEW

File: devito/types/dimension.py [hunk 1]
```diff
@@ -892,6 +895,18 @@ def condition(self):
...
+    @property
+    def fact_index(self):
+        if self.condition is None or self._factor is None:
+            return self.index
+
+        # This is the corner case where both a condition and a factor are provided
+        # the index will need to be `self.parent - min(self.condition)` to avoid
+        # shifted indexing. E.g if you have `factor=2` and `condition=Ge(time, 10)`
+        # then the lowered index needs to be `(time - 10)/ 2`
+        ltkn = relational_min(self.condition, self.parent)
```
Comment (FabioLuporini) [inline devito/types/dimension.py:907]: ltkn stands for "left thickness" -- is it really the name u wanna use here?

(nitpicking)

File: devito/types/dimension.py [hunk 2]
```diff
@@ -927,7 +942,7 @@ def _arg_defaults(self, _min=None, size=None, alias=None):
         # `factor` endpoints are legal, so we return them all. It's then
         # up to the caller to decide which one to pick upon reduction
         dim = alias or self
-        if dim._factor is None or size is None:
+        if dim.condition is not None or size is None:
```
Comment (FabioLuporini) [inline devito/types/dimension.py:945]: so we don't check _factor anymore here?

File: devito/types/dimension.py (hunk 1 above)
Comment (mloubout) [reply devito/types/dimension.py:907]: Yeah that's intentional that the actual left thickness that is required to be shifted by.

File: devito/types/dimension.py (hunk 2 above)
Comment (mloubout) [reply devito/types/dimension.py:945]: Yeah, basically if there is a condition then you can't safely use the factor to infer arg bounds

File: devito/types/dimension.py [hunk 3]
```diff
@@ -892,6 +895,18 @@ def condition(self):
     def indirect(self):
         return self._indirect
 
+    @property
+    def fact_index(self):
```
Comment (FabioLuporini) [inline devito/types/dimension.py:899]: can you not move this whole logic into `__init_finalize__` and set index to the proper value from the get go?

File: devito/ir/equations/equation.py [hunk 4]
```diff
@@ -188,12 +188,19 @@ def __new__(cls, *args, **kwargs):
         for d in ordering:
             if not d.is_Conditional:
                 continue
-            if d.condition is None:
+            if d._factor is not None:
```
Comment (FabioLuporini) [inline devito/ir/equations/equation.py:191]: why not `d.factor`?

**Review by FabioLuporini — APPROVED** [review]

File: devito/ir/equations/equation.py (hunk 4 above)
Comment (mloubout) [reply devito/ir/equations/equation.py:191]: `d.factor` is never None it returns `1` if there is no factor for various reasons

File: devito/types/relational.py [hunk 5]
```diff
@@ -208,3 +209,44 @@ def __new__(cls, lhs, rhs=0, subdomain=None, **kwargs):
 
 ops = {Ge: Lt, Gt: Le, Le: Gt, Lt: Ge}
 rev = {Ge: Le, Gt: Lt, Lt: Gt, Le: Ge}
+
+
+def relational_min(expr, s):
```
Comment (georgebisbas) [inline devito/types/relational.py:214]: Any chance you may reuse:

`def evalrel(func=min, input=None, assumptions=None):`

have not looks closely but maybe we can find some common way

File: devito/types/relational.py [hunk 6]
```diff
@@ -208,3 +209,44 @@ def __new__(cls, lhs, rhs=0, subdomain=None, **kwargs):
...
+    """
+    Infer the minimum valid value for symbol `s` in the expression `expr`.
+    For example
+        - if `expr` is `s < 10`, then the minimum valid value for `s` is 0
+        - if `expr` is `s >= 10`, then the minimum valid value for `s` is 10
```
Comment (georgebisbas) [inline devito/types/relational.py:219]: or at least this rang a bell when reading the docstring here

File: devito/types/relational.py (hunk 5 above)
Comment (mloubout) [reply devito/types/relational.py:214]: No `evalrel` simplifies/... a set of relations, this does not evaluate the condition to some minimal case, it finds the smallest. value for symbol `s` that satisfies `expr` in order to get a index bound

**Review by georgebisbas — APPROVED** [review]

**Pull request merged**